packages/web/src/components/ServerInject.tsx

```diff
@@ -0,0 +1,67 @@
+import React, { Fragment, ReactNode, useContext } from 'react'
+
+/**
+ *
+ * Inspired by Next's useServerInsertedHTML, originally designed for CSS-in-JS
+ * for now it seems the only way to inject html with streaming is to use a context
+ *
+ * We use this for <head> tags, and for apollo cache hydration
+ *
+ * Until https://github.com/reactjs/rfcs/pull/219 makes it into react
+ *
+ */
+
+export type RenderCallback = () => ReactNode
+
+export const ServerHtmlContext = React.createContext<
+  ((things: RenderCallback) => void) | null
+>(null)
+
+/**
+ *
+ *  Use this factory, once per request.
+ *  This is to ensure that injectionState is isolated to the request
+ *  and not shared between requests.
+ */
+export const createInjector = () => {
+  const injectionState: Set<RenderCallback> = new Set([])
+
+  const injectToPage = (renderCallback: RenderCallback) => {
+    injectionState.add(renderCallback)
+  }
+
+  return { injectToPage, injectionState }
+}
+
+// @NOTE do not instatiate the provider value here, so that we can ensure
+// context isolation. This is done in streamHelpers currently,
+// using the createInjector factory, once per request
+export const ServerHtmlProvider = ServerHtmlContext.Provider
+
+export const ServerInjectedHtml = ({
+  injectionState,
+}: {
+  injectionState: Set<RenderCallback>
+}) => {
+  const serverInsertedHtml = []
+  for (const callback of injectionState) {
+    serverInsertedHtml.push(callback())
+
+    // Remove it from the set so its not rendered again
+    injectionState.delete(callback)
+  }
+
+  return serverInsertedHtml.map((html, i) => {
+    return <Fragment key={`rwjs-server-injected-${i}`}>{html}</Fragment>
```

Comment:
Don't have full context here, assuming it's ok/you guys worked it out, but using the index of an item in an array isn't recommended (see the pitfall here: https://react.dev/learn/rendering-lists#why-does-react-need-keys). What's the nuance here? Maybe document it here in a comment too

Reply:
There isn't really any nuance. I think it doesn't matter if we used i here anyway - but I've changed it now to have unique-id + key. The unique ids should never match.